rust/lance-linalg/src/simd/f16.c

```diff
@@ -55,3 +56,26 @@ float l2_f16(const _Float16 *x, const _Float16 *y, uint32_t dimension) {
   }
   return (float) sum;
 }
+
+/// @brief Fast cosine function, that assumes that the norm of the first vector is already known.
+/// @param x A f16 vector
+/// @param x_norm Norm of the first vector
+/// @param y A f16 vector
+/// @param dimension The dimension of the vectors
+/// @return The cosine distance of the two vectors.
+float cosine_fast_f16(const _Float16 *x, const float x_norm, const _Float16 *y, uint32_t dimension) {
+  _Float16 xy = 0.0;
+  _Float16 y_sq = 0.0;
+
+#pragma clang loop unroll(enable) interleave(enable) vectorize_width(32)
+  for (uint32_t i = 0; i < dimension; i++) {
+    const _Float16 yi = y[i];
+    xy += x[i] * yi;
```

Comment:
Have we verified that compiling with supported cpu target,  avx512fp16 instructions will be used?

https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point

Reply:
Using https://godbolt.org/ I can confirm that we get avx512 fp16 instructions used here.